### PR TITLE
Add `@types/bun`

### PR DESCRIPTION
**Pull request by colinhacks**
[Edit: Dec 14th]

Per the discussion, this is now a shim that loads `bun-types` via tripleslash. This PR adds `bun-types` to the list of allowable dependencies: https://github.com/microsoft/DefinitelyTyped-tools/pull/881

```ts
/// <reference types="bun-types" />
```

This package contains no declarations or tests. We've roughly replicated the DT linting setup in our own repo here: https://github.com/oven-sh/bun/pull/7670 We run typechecking on a suite of 40+ test files. We've rewritten `bun-types` to avoid any type conflicts with `lib.dom.d.ts` or `@types/node`. 

Versioning scheme:
- Once this PR lands, we'll merge https://github.com/oven-sh/bun/pull/7670
- Versions of `@types/bun` will have a dependency on `bun-types` pinned to a specific version
- When new versions of `bun-types` land, a member of the Bun team (or our CI) will submit a PR to bump the dependency in `@types/bun` 

---

[Outdated]

This is a fresh PR adding `@types/bun`. It now incorporates `@types/node` as a dependency instead of redeclaring those modules. 

Previous attempt: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67275. You can see the discussion of the disabled lint rules here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67275/files#r1380753400

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/bun/sqlite.d.ts
1. types/bun/test.d.ts
1. types/bun/bun.d.ts
1. types/bun/ffi.d.ts
1. types/bun/jsc.d.ts

<details>
<summary>as well as these 8 other files...</summary>
<p>types/bun/globals.d.ts, types/bun/test/globals.test.ts, types/bun/test/ffi.test.ts, types/bun/test/array.test.ts, types/bun/test/index.test.ts, types/bun/test/console.test.ts, types/bun/tsconfig.json, types/bun/test/wasm.test.ts</p>
</details>


Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against fae2c69b37acb4e4af361d8282ddec335c34be99
</p>

File: types/bun/package.json
Comment (eps1lon) [inline types/bun/package.json:9]: Is this actually how Bun works or is this just for convenience? Because now you're saying that any Bun version has the latest APIs of Node.js

File: types/bun/package.json
Comment (peterblazejewicz) [inline types/bun/package.json:16]: all those users would receive notification, is that ok for them?
Could we one get their ack here? (or should you just give them credits)
https://github.com/DefinitelyTyped/DefinitelyTyped#definition-owners

Comment (colinhacks) [reply types/bun/package.json:16]: Yes, those are all Bun team members. We want to be able to merge definition PRs quickly so some redundancy here is fine.

File: types/bun/package.json
Comment (colinhacks) [reply types/bun/package.json:9]: By and large, it's how Bun works, but it's true that this is partially aspirational. My previous PR re-declared types for all built-in Node.js modules and we received the opposite feedback (too redundant). There's tradeoffs with both but ultimately Bun aims to be fully Node compatible and IMO our type declarations are not the right place to indicate  deviations from that goal. We document our compatibility here: https://bun.sh/docs/runtime/nodejs-apis

Comment (colinhacks) [reply types/bun/package.json:9]: Also, if we redeclare these modules, it is virtually impossible to avoid compilation errors if a user has both `@types/bun` and `@types/node` installed simultaneously, which will inevitably happen very frequently. Jake Bailey  rightly expressed this concern here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67275#issuecomment-1793353029

Comment (peterblazejewicz) [reply types/bun/package.json:9]: This seems acceptable. (afaik all DT packages do this that way, if used)

**Comment by peterblazejewicz** [comment]
/cc @jakebailey as you spent some time and had some input on previous effort
(looks good, will do good things when accepted and used as prescribed IMO)

File: types/bun/tslint.json
Comment (jakebailey) [inline types/bun/tslint.json:6]: No need for this file anymore; these rules are not enabled (and tslint is one PR from going away).

File: types/bun/README.md
Comment (jakebailey) [inline types/bun/README.md:25]: This is more canonically written `"bun"`.

Comment (jakebailey) [reply types/bun/README.md:25]: But, I thought half of the point was to not have to manually specify anything?

Comment (paperdave) [reply types/bun/README.md:25]: yeah, i think this entire "usage" section should be removed

**Comment by colinhacks** [comment]
> I would have just reopened #67275 on request as we just lost all of the discussion ☹️

Apologies

> Not repeating the types is good, so long as you're happy with the downside of over-declaring the API.

Agreed, it's a good tradeoff. 👍
 
> * for the auto-inclusion (which at this point, IMO is harmful)

This is still a point of concern for us, yeah. I understand the concerns with auto-inclusion but it's an uphill battle that we'd rather not fight. 

> I am still not entirely certain how valuable it is having this on DT when `bun-types` is the same thing with a release schedule and versioning scheme you control, though.

I just had a thought — would it be totally out of pocket for us to make `@types/bun` a shell package with a dependency on `bun-types`? Then use `/// <reference types="bun-types" />` to register our globals? I should note that after this PR gets merged we're planning to keep our types in the oven-sh/bun repo and set up CI to automatically submit PRs to DT. But that's a lot of moving parts, and we'd have to replicate the linting setup precisely...it all seems pretty fragile. The shell package approach would be a happy middle ground, but if there's a hard policy against the shell package thing, I'd understand.

**Comment by jakebailey** [comment]
> I understand the concerns with auto-inclusion but it's an uphill battle that we'd rather not fight. We've gotten literally hundreds of support requests from users about TS errors in VS Code because they don't know about `"types"` and are used to auto-inclusion.

I'm honestly quite surprised by this. But, maybe I've avoided this pain thanks to always using `bun --init`, which nicely configures everything for me. How are people hitting this? Migrating existing projects? Something else?

> This is still a point of concern for us, yeah. I should note that after this PR gets merged we're planning to keep our types in the oven-sh/bun repo and set up CI to auto-submit PRs to DT. We'll have to replicate the linting configuration exactly and it'll probably be fiddly, but its just too difficult to try to synchronize PR merging across two repos.

Some packages do this. The most immediate example is `@types/three`, though there are others. Many are generated code.

The main problem with this approach is that people _will_ send PRs to change things, and maintainers will fix things that have to do with DT infra changes or to fix things for upcoming versions of TypeScript. Often these changes get forgotten on the next import, or cause a break because they weren't upstreamed back again (e.g. fail on TS@next or something, confusing the maintainer).

It's kinda weird, because at that point it's not really a contributable project, but just a publishing mechanism that uses our resources.

> I just had a thought — would it be totally out of pocket for us to make `@types/bun` a shell package with a dependency on `bun-types`? Then use `/// <reference types="bun-types" />` to register our globals? Seems like a happy middle ground, but if there's a hard policy against the shell package thing, I'd understand.

Such a thing isn't new, no. The main downside is the lack of control users may have over versioning as the actual package doing the work is a transitive dep. That makes it hard to handle unless the DT package also pins a version, in which case there's not too much difference between the two. The closest example would be `ramda` or any of the `gapi.*` packages, and to some extent, `jest`.

I think I'd personally prefer it being on DT in full or not being on DT at all, as opposed to it being a wrapper.

I am also just one voice; @sandersn and @andrewbranch may have some other thoughts here.

**Comment by colinhacks** [comment]
> How are people hitting this? Migrating existing projects? Something else?

Precisely, yeah. They have an existing project, start using the Bun runtime to run their dev server, decide to use a `Bun` API, install `bun-types`, then get confused when `Bun` still isn't recognized.


> Such a thing isn't new, no. The main downside is the lack of control users may have over versioning as the actual package doing the work is a transitive dep. That makes it hard to handle unless the DT package also pins a version, in which case there's not too much difference between the two. The closest example would be ramda or any of the gapi.* packages, and to some extent, jest.

In the absence of strong objection, this seems to be the best avenue for us. We can bump the DT package every time we do a minor release, but this also lets us publish bugfixes immediately, and supports auto-inclusion.

**Comment by andrewbranch** [comment]
My personal philosophy is that the purpose of DefinitelyTyped is to allow the community to create and collaborate on things that are valuable to the community, `@types` publishing exists to remove friction for users getting the fruits of that collaboration, and auto-inclusion of `@types` (or really any special treatment) by the compiler was a mistake. So on principle, I don’t _like_ to see systems and practices set up specifically to take advantage of this mistake while also moving the center of collaboration for these types away from DT.

However, that’s just the natural result of the incentives we created, and there’s no policy against it. I would rather see the shell package than for you all to have to do a ridiculous amount of work to automate PRs and set up DT infrastructure on your own codebase. There’s no point in creating a review bottleneck here on development that’s happening somewhere else completely. I guess just be aware that in the long run, I would really like to re-examine the incentive structure we have here, including `@types` auto-inclusion in TS.

**Comment by jakebailey** [comment]
I am in agreement with the above. The shim seems fine, we'll just have to figure out versioning.

**Comment by sandersn** [comment]
Same here.

File: types/bun/index.d.ts
Comment (jakebailey) [inline types/bun/index.d.ts:1]: This header should be removed.

File: types/bun/.eslintrc.json
Comment (jakebailey) [inline types/bun/.eslintrc.json:1]: Is this config still needed?

Comment (colinhacks) [reply types/bun/.eslintrc.json]: Removed

File: types/bun/index.d.ts
Comment (colinhacks) [reply types/bun/index.d.ts]: Removed

File: types/bun/package.json
Comment (jakebailey) [inline types/bun/package.json:9]: Just to be sure, is the intent that you'll send a version bump each time? Or would unpinned be better?

Comment (colinhacks) [reply types/bun/package.json:9]: Without pinning, it's conceivable that some package managers will install an older version of `bun-types` from cache. Perhaps ironically this is the kind of optimization `bun install` is likely to do. 

It's not a big hassle for us to bump the dependency version, and it's easy to automate for us. Also - changing this value in the package.json is a perfect way for us to trigger DT's publishing CI.

Comment (jakebailey) [reply types/bun/package.json:9]: Alrighty, just know that there may be some delay depending on reviewers, though you may be able to self merge until you're popular. (Of course I can't not look at things I'm pinged on so... that's one way to guarantee a review 😅)

**Comment by jakebailey** [comment]
I've just sent https://github.com/microsoft/TypeScript/pull/56795 which special cases the error message for a missing `Bun` variable, similar to what we have for jquery/node/mocha/jest.

**Comment by jakebailey** [comment]
I think that this is all good at this point. Any other comments?

File: types/bun/package.json
Comment (colinhacks) [reply types/bun/package.json:9]: Ah I did think we'd be able to self-merge going forward. Let's still leave it as is and see if it becomes burdensome for either us or the DT maintainers. Curious what the cutoff for "popular" is - `bun-types` gets just over 100k weekly downloads: https://www.npmjs.com/package/bun-types

> Of course I can't not look at things I'm pinged on

You are the platonic ideal of a maintainer 🫶 Take some breaks 😅

**Comment by colinhacks** [comment]
> I've just sent https://github.com/microsoft/TypeScript/pull/56795 which special cases the error message for a missing Bun variable, similar to what we have for jquery/node/mocha/jest.

Thank you! 🙏 Left a tiny nitpick

> I think that this is all good at this point. Any other comments?

Ready to merge when you are 🚀

**Review by jakebailey — APPROVED** [review]

**Comment by colinhacks** [comment]
Ready to merge

**Pull request merged**
